src/components/fileviewer.js

```diff
@@ -15,88 +15,93 @@ export default class FileViewerContainer extends Component {
     }
-    this.revision = parsedQuery.revision
-    this.path = parsedQuery.path
+    this.revision = parsedQuery.revision;
+    this.path = parsedQuery.path;
+  }
 
+  componentDidMount() {
     /* Fetch source code from hg */
     FetchAPI.getRawFile(this.revision, this.path)
-      .then(text => this.setState({ parsedFile: text.split("\n") }))
+      .then(text => this.setState({ parsedFile: text.split('\n') }))
     ;
 
     /* Fetch coverages from ActiveData */
     FetchAPI.query({
-      "from": "coverage",
-      "where": {
-        "and": [
-          {"eq": {"source.file.name": `${this.path}`}},
-          {"eq": {"repo.changeset.id12": `${this.revision}`}}
-        ]
+      from: 'coverage',
+      where: {
+        and: [
+          { eq: { 'source.file.name': `${this.path}` } },
+          { eq: { 'repo.changeset.id12': `${this.revision}` } },
+        ],
       },
-      "limit": 1000,
-      "format": "list"
+      limit: 1000,
+      format: 'list',
     })
-    .then(data => {
-      console.log(data);
-      this.setState({coverage: data});
-      this.parseTestsCoverage(data.data);
-    });
+      .then((data) => {
+        this.parseCoverage(data.data);
+      });
   }
 
-  /* Parse data returns from ActiveData */
-  parseTestsCoverage(data) {
-    console.log(data.length);
-    var stat = [];
+  setSelectedLine(selectedLineNumber) {
+    this.setState({ selectedLine: selectedLineNumber });
+  }
 
-    // initial list for each coverable line
-    if (data.length > 0) {
-      const fileCov = data[0].source.file;
-      fileCov.covered.concat(fileCov.uncovered).forEach(line => {
-        stat[line] = [];
+  /* Parse coverage data */
+  parseCoverage(data) {
+    const covered = [];
+    const uncovered = [];
+    const testsPerHitLine = [];
+    const testsPerMissLine = [];
+
+    data.forEach((d) => {
+      d.source.file.covered.forEach((line) => {
+        covered.push(line);
+        if (!testsPerHitLine[line]) {
+          testsPerHitLine[line] = [];
+        }
+        testsPerHitLine[line].push(d);
       });
-    }
-
-    data.forEach(d => {
-      d.source.file.covered.forEach(line => {
-        stat[line].push(d);
+      d.source.file.uncovered.forEach((line) => {
+        uncovered.push(line);
+        if (!testsPerMissLine[line]) {
+          testsPerMissLine[line] = [];
+        } else {
+          testsPerMissLine[line].push(d);
+        }
       });
     });
 
-    console.log(stat);
-    this.setState({testsPerLines: stat});
-  }
-
-  /* handle fileviewer's line onclick event */
-  setSelectedLine(lineNumber) {
-    console.log(lineNumber);
-    this.setState({selectedLine: lineNumber});
-    // if (this.state.selectedLine == lineNumber) {
-    //   this.setState({selectedLine: null})
-    // } else {
-    //   this.setState({selectedLine: lineNumber});
-    // }
+    this.setState({
+      coverage: {
+        coveredLines: _.uniq(covered),
+        uncoveredLines: _.uniq(uncovered),
```

Comment:
How about we first get which lines are coverable, and then use ```coverableLines``` - ```coverdLines``` to get the ```uncoveredLines```?

Reply:
yes, there used to be a subtraction after this step.  Probably lost in the refactor.

Reply:
I left the subtraction step in CoveragePercentageViewer but I guess it is better to have it up here